org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,1296 @@
+        // isn't called. In which case, we'll need to explicitly query the
+        // list off the ssl_fd.
+        if (enabled_ciphers == null && ssl_fd == null) {
+            queryEnabledCipherSuites();
+        }
+
+        if (enabled_ciphers == null) {
+            // TODO: Query default ciphersuites here.
+            throw new RuntimeException("Unable to query enabled ciphers off of empty ssl_fd.");
+        }
+
+        // Use JSSParameters to do the heavy lifting of converting our list
+        // of cipher suites to an array of Strings.
+        JSSParameters parser = new JSSParameters();
+        parser.setCipherSuites(enabled_ciphers);
+        return parser.getCipherSuites();
+    }
+
+    private void queryEnabledProtocols() {
+        logger.debug("JSSEngine: queryEnabledProtocols()");
+        SSLVersionRange vrange = null;
+        try {
+            vrange = SSL.VersionRangeGet(ssl_fd);
+        } catch (Exception e) {
+            // This shouldn't happen unless the PRFDProxy is null.
+            throw new RuntimeException("JSSEngine.queryEnabledProtocols() Unexpected failure: " + e.getMessage(), e);
+        }
+
+        if (vrange == null) {
+            // Again; this shouldn't happen as the vrange should always
+            // be created by VersionRangeGet(...).
+            throw new RuntimeException("JSSEngine.queryEnabledProtocols() - null protocol range");
+        }
+
+        min_protocol = vrange.getMinVersion();
+        max_protocol = vrange.getMaxVersion();
+    }
+
+    public String[] getEnabledProtocols() {
+        logger.debug("JSSEngine: getEnabledProtocols()");
+
+        if ((min_protocol == null || max_protocol == null) && ssl_fd != null) {
+            queryEnabledProtocols();
+        }
+
+        if (min_protocol == null || max_protocol == null) {
+            // TODO: Query default ciphersuites here.
+            throw new RuntimeException("JSSEngine.getEnabledProtocls() - Unable to query enabled protocols off of empty ssl_fd.");
+        }
+
+        // Use JSSParameters to do the heavy lifting of converting the
+        // SSLVersionRange to a list of JDK-conforming Strings.
+        JSSParameters parser = new JSSParameters();
+        parser.setProtocols(min_protocol, max_protocol);
+        return parser.getProtocols();
+    }
+
+    public boolean getEnableSessionCreation() {
+        logger.debug("JSSEngine: getEnableSessionCreation() - not implemented");
+        return true;
+    }
+
+    public SSLEngineResult.HandshakeStatus getHandshakeStatus() {
+        logger.debug("JSSEngine: getHandshakeStatus()");
+        try {
+            updateHandshakeState();
+        } catch (SSLException se) {
+            // Do nothing -- updateHandshakeStatus() should throw the same
+            // exception again, so it is safe to discard it. We also can't
+            // just pass the SSLException to the caller of this method, as
+            // they aren't expecting it.
```

Comment:
If the error might affect subsequent operations, we can always wrap it with a RuntimeException. If we can safely ignore it, we probably should at least log a warning message. A blank catch block generally will make it harder to troubleshoot.

Reply:
It won't affect any operations other than `wrap(...)` or `unwrap(...)`, which are already expecting an exception.... ...and which this very same exception will be raised from.

The issue is there's a bit of a dance here. This is meant to be a "static" getter, probably just returning the value of a private variable. However, we use this as an opportunity to step NSS's internal state machine, in case we're "stuck". A typical loop looks like:

```java
while (eng.getHandshakeStatus() != FINISHED) {
    try {
        eng.wrap(...);
    } // handle
    try {
        eng.unwrap(...);
    } // handle
}
```

So clearly the caller isn't expecting `getHandshakeStatus()` to throw an exception.

So I've made `updateHandshakeState()` store whatever exception it throws for the next time it is called. After the first handshake exception, the TLS protocol is sunk, so you really just write an alert to the write and give up.